reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java

```diff
@@ -0,0 +1,829 @@
+						Borrower borrower = borrowers.pollFirst();
+						if (borrower == null) {
+							continue;
+						}
+						if (isDisposed()) {
+							borrower.fail(new PoolShutdownException());
+							return;
+						}
+						borrower.stopPendingCountdown();
+						Mono<Connection> allocator = poolConfig.allocator();
+						Mono<Connection> primary =
+								allocator.doOnEach(sig -> {
+								             if (sig.isOnNext()) {
+								                 Connection newInstance = sig.get();
+								                 assert newInstance != null;
+								                 Slot newSlot = new Slot(this, newInstance);
+								                 newSlot.activate();
+								                 ACQUIRED.incrementAndGet(this);
+								                 newSlot.incrementConcurrency();
+								                 newSlot.deactivate();
+								                 borrower.deliver(new Http2PooledRef(newSlot));
+								             }
+								             else if (sig.isOnError()) {
+								                 Throwable error = sig.getThrowable();
+								                 assert error != null;
+								                 poolConfig.allocationStrategy().returnPermits(1);
+								                 borrower.fail(error);
+								             }
+								         })
+								         .contextWrite(borrower.currentContext());
+
+						if (permits == 1) {
+							primary.subscribe(alreadyPropagated -> {}, alreadyPropagatedOrLogged -> drain(), this::drain);
+						}
+						else {
+							int toWarmup = permits - 1;
+							if (log.isDebugEnabled()) {
+								log.debug("should warm up {} extra connections", toWarmup);
+							}
+
+							Flux<Void> warmupFlux =
+									Flux.range(1, toWarmup)
+									    .flatMap(i -> allocator.flatMap(poolable -> {
+									        if (log.isDebugEnabled()) {
+									            log.debug("warmed up extra connections {}/{}", i, toWarmup);
+									        }
+									        Slot newSlot = new Slot(this, poolable);
+									        if (!offerSlot(newSlot)) {
+									            poolConfig.allocationStrategy().returnPermits(1);
+									            releaseConnection(poolable);
+									            return Mono.<Void>empty();
+									        }
+									        return Mono.empty();
+									    })
+									    .onErrorResume(warmupError -> {
+									        if (log.isDebugEnabled()) {
+									            log.debug("failed to warm up extra connections {}/{}: {}", i, toWarmup, warmupError.toString());
+									        }
+									        poolConfig.allocationStrategy().returnPermits(1);
+									        return Mono.empty();
+									    }));
+
+							primary.onErrorResume(e -> Mono.empty())
+							       .thenMany(warmupFlux)
+							       .subscribe(aVoid -> {}, alreadyPropagatedOrLogged -> drain(), this::drain);
+						}
+					}
+				}
+			}
+
+			if (WIP.decrementAndGet(this) == 0) {
+				recordInteractionTimestamp();
+				break;
+			}
+		}
+	}
+
+	@Nullable
+	Slot findConnection(ConcurrentLinkedQueue<Slot> resources) {
```

Comment:
would it make sense for `findConnection` to kinda "reserve" a `Slot` it found with `activate()`? I'm just wondering if by the time we `activate()` it in the drain loop, the various criteria checked below could have changed, and if so does activating early help... 🤔

Reply:
we have 3 main things checked in `findConnection`

1. channel has been closed
  * if the channel is closed, this cannot be changed
  * if the channel is opened and then closed at some time - this will fail on opening a stream and I don't see how we can prevent this with `activate()`
2. channel has reached max life time
  * if the channel reached max life time, this cannot be changed
  * if the channel can still be used and then later it reaches max life time - then this will be the last stream that will be opened on this channel and then the channel will be closed. I don't see how we can prevent this with `activate()`.
3. channel has reached max number of active streams - the channel will not be used and will stay in the pool. When some active stream finishes its work and is closed we will decrease the concurrency and drain will be invoked (destroyPoolable) at that time the channel will be able to be used again.  I don't see how we can prevent this with `activate()`.